core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -227,48 +276,131 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
     // This message is received for one of these reasons:
     // 1. Container errored while resuming a warm container, could not process the job, and sent the job back
     // 2. The container aged, is destroying itself, and was assigned a job which it had to send back
     // 3. The container aged and is destroying itself
+    // 4. The container is paused, and being removed to make room for new containers
     // Update the free/busy lists but no message is sent to the feed since there is no change in capacity yet
     case RescheduleJob =>
+      resourceManager.releaseReservation(sender())
       freePool = freePool - sender()
       busyPool = busyPool - sender()
+      updateUnused()
+    case InitPrewarms =>
+      initPrewarms()
+    case ResourceUpdate => //we may have more resources - either process the buffer or request another feed message
+      processBuffer()
+    case ContainerStarted => //only used for receiving post-start from cold container
+      //stop tracking via reserved
+      resourceManager.releaseReservation(sender())
+    case NeedResources(size: ByteSize) => //this is the inverse of NeedWork
+      MetricEmitter.emitCounterMetric(LoggingMarkers.CONTAINER_POOL_RESOURCE_ERROR)
+      //we probably are here because resources were not available even though we thought they were,
+      //so preemptively request more
+      resourceManager.requestSpace(size)
+    case ReleaseFree(refs) =>
+      logging.info(this, s"pool is trying to release ${refs.size} containers by request of invoker")
+      //remove each ref, IFF it is still not in use, and has not been used since the removal was requested
+      ContainerPool.findIdlesToRemove(freePool, refs).foreach(removeContainer)
+    case EmitMetrics =>
+      logging.info(
+        this,
+        s"metrics invoker (self) has ${runBuffer.size} buffered (${runBuffer.map(_.action.limits.memory.megabytes).sum}MB)")
+
+      MetricEmitter.emitGaugeMetric(LoggingMarkers.CONTAINER_POOL_RUNBUFFER_COUNT, runBuffer.size)
+      MetricEmitter.emitGaugeMetric(
+        LoggingMarkers.CONTAINER_POOL_RUNBUFFER_SIZE,
+        runBuffer.map(_.action.limits.memory.megabytes).sum)
+      val containersInUse = inUse
+      MetricEmitter.emitGaugeMetric(LoggingMarkers.CONTAINER_POOL_ACTIVE_COUNT, containersInUse.size)
+      MetricEmitter.emitGaugeMetric(
+        LoggingMarkers.CONTAINER_POOL_ACTIVE_SIZE,
+        containersInUse.map(_._2.memoryLimit.toMB).sum)
+      MetricEmitter.emitGaugeMetric(LoggingMarkers.CONTAINER_POOL_PREWARM_COUNT, prewarmedPool.size)
+      MetricEmitter.emitGaugeMetric(
+        LoggingMarkers.CONTAINER_POOL_PREWARM_SIZE,
+        prewarmedPool.map(_._2.memoryLimit.toMB).sum)
+  }
+
+  /** Buffer processing in cluster managed resources means to send the first item in runBuffer;
+   *  In non-clustered case, it means signalling MessageFeed (since runBuffer is processed in tight loop).
+   * */
+  def processBuffer() = {
+    if (poolConfig.clusterManagedResources) {
```

Comment:
Any reason for this logic to be restricted to cluster managed resources flow only (unless intention is to be strictly backward compatible with previous flow for non cluster manage case)? 

So far the job from `runBuffer` was sent back only when current job was sent for processing. For `NeedWork` and `ContainerRemoved` case `runBuffer` was not touched. But logically it should be ok to first check for `runBuffer` and then only ask for more work

Reply:
Agreed - removed the restriction so that processBuffer() is called for all (cluster managed, and not).